### PR TITLE
OTLP LogExporter to special case {OriginalFormat} to populate body

### PR DESCRIPTION
**Pull request by cijothomas**
This PR modifies OTLP LogExporter to:
"specially" treat {OriginalFormat} key from `LogRecord.StateValues`, and populate Body from its value, if `LogRecord.FormattedMessage` is not available.

Why: 
[Official docs on logging](https://docs.microsoft.com/dotnet/core/extensions/logging?tabs=command-line#log-message-template) recommends (or rather *only talks about*), using the [extensions methods](https://docs.microsoft.com/dotnet/api/microsoft.extensions.logging.loggerextensions?view=dotnet-plat-ext-6.0) to do logging, which would [result in storing the "template" as {OriginalFormat}](https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Extensions.Logging.Abstractions/src/LogValuesFormatter.cs#L198)

Since this is the most common scenario, I am proposing to "special" case this, to populate body from. (If user opted-in for `IncludeFormattedMessage=true`, then this special casing is not applied, and {OriginalFormat} get stores as just any other attribute, and Body will be the `LogRecord.FormattedMessage`. Okay to change that behavior if there is agreement.)

### CODE REVIEW

File: test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpLogExporterTests.cs [hunk 1]
```diff
@@ -225,28 +225,82 @@ public void CheckToOtlpLogRecordSeverityLevelAndText(LogLevel logLevel)
...
 
             Assert.NotNull(otlpLogRecord);
-            Assert.Equal(logRecord.FormattedMessage, otlpLogRecord.Body.StringValue);
+            if (includeFormattedMessage)
+            {
+                Assert.Equal(logRecord.FormattedMessage, otlpLogRecord.Body.StringValue);
+            }
+            else
+            {
+                Assert.Equal("OpenTelemetry {Greeting} {Subject}!", otlpLogRecord.Body.StringValue);
```
Comment (cijothomas) [inline test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpLogExporterTests.cs:260]: This is the "special" treatment of {OriginalFormat} key.

File: test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpLogExporterTests.cs [hunk 2]
```diff
@@ -225,28 +225,82 @@ public void CheckToOtlpLogRecordSeverityLevelAndText(LogLevel logLevel)
...
+
+            Assert.NotNull(otlpLogRecord);
+
+            // There is no formatter, so no way to populate Body.
+            // Exporter won't even attempt to do ToString() on State.
+            if (includeFormattedMessage)
+            {
+                Assert.Null(otlpLogRecord.Body);
+            }
+            else
```
Comment (cijothomas) [inline test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpLogExporterTests.cs:300]: intentionally making the same Assert on If..Else, to make it easy for readers :)

**Review by jviau — APPROVED** [review]
LGTM

File: test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpLogExporterTests.cs [hunk 3]
```diff
@@ -225,28 +225,82 @@ public void CheckToOtlpLogRecordSeverityLevelAndText(LogLevel logLevel)
...
         {
             var logRecords = new List<LogRecord>();
             using var loggerFactory = LoggerFactory.Create(builder =>
             {
                 builder.AddOpenTelemetry(options =>
                 {
                     options.AddInMemoryExporter(logRecords);
-                    options.IncludeFormattedMessage = true;
+                    options.IncludeFormattedMessage = includeFormattedMessage;
+                    options.ParseStateValues = true;
```
Comment (alanwest) [inline test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpLogExporterTests.cs:240]: Do we need a test for when both IncludeFormattedMessage and ParseStateValues is null then Body will be null?

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/LogRecordExtensions.cs [hunk 4]
```diff
@@ -76,17 +76,26 @@ internal static OtlpLogs.LogRecord ToOtlpLog(this LogRecord logRecord)
...
+                    bodyPopulatedFromFormattedMessage = true;
                 }
 
                 if (logRecord.StateValues != null)
                 {
                     foreach (var stateValue in logRecord.StateValues)
                     {
-                        var otlpAttribute = stateValue.ToOtlpAttribute();
-                        otlpLogRecord.Attributes.Add(otlpAttribute);
+                        if (stateValue.Key.Equals("{OriginalFormat}") && !bodyPopulatedFromFormattedMessage)
```
Comment (alanwest) [inline src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/LogRecordExtensions.cs:90]: A comment here referencing this is the behavior when using the extension methods will help my memory 6 months from now 😄

File: test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpLogExporterTests.cs (hunk 3 above)
Comment (cijothomas) [reply test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpLogExporterTests.cs:240]: yes surely. (There are lot of corner cases as well)
I'll iterate to add more tests.

**Review by alanwest — APPROVED** [review]
LGTM

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/CHANGELOG.md [hunk 5]
```diff
@@ -2,9 +2,12 @@
 
 ## Unreleased
 
-* LogExporter to correcly map Severity to OTLP.
+* LogExporter to correctly map Severity to OTLP.
   ([#3177](https://github.com/open-telemetry/opentelemetry-dotnet/pull/3177))
 
+* LogExporter to special case {OriginalFormat} to populate
```
Comment (reyang) [inline src/OpenTelemetry.Exporter.OpenTelemetryProtocol/CHANGELOG.md:8]:
```suggestion
* LogExporter to special case '{OriginalFormat}' to populate
```

**Review by reyang — APPROVED** [review]
LGTM.

**Review by CodeBlanch — APPROVED** [review]
LGTM
